src/Network/Transport/TCP.hs

```diff
@@ -828,17 +834,28 @@ apiCloseEndPoint transport evs ourEndPoint =
             putMVar resolved ()
             return closed
           RemoteEndPointValid vst -> do
+            -- Schedule an action to send a CloseEndPoint message and then
+            -- wait for the socket to actually close (meaning that this
+            -- end point is no longer receiving from it).
+            -- Since we replace the state in this MVar with 'closed', it's
+            -- guaranteed that no other actions will be scheduled after this
+            -- one.
             sched theirEndPoint $ do
               void $ tryIO $ sendOn vst [ encodeInt32 CloseEndPoint ]
               -- Release probing resources if probing.
               forM_ (remoteProbing vst) id
-              tryCloseSocket (remoteSocket vst)
```

Comment:
I fathom this used to cause the thread running `handleIncomingMessages` to terminate with an exception when reading from the socket. Now that this is removed, how does the implementation make sure that the thread terminates in a timely fashion?

It looks as if now it waits for the remote side to close the socket first, but if there is a network failure, the socket may be closed remotely, and yet the local transport may not learn about it.